src/Catalog/Persistence/AzureStorage.cs

```diff
@@ -379,21 +380,57 @@ public override Uri GetUri(string name)
 
         public override async Task<bool> AreSynchronized(Uri firstResourceUri, Uri secondResourceUri)
         {
-            var destination = GetBlockBlobReference(GetName(secondResourceUri));
             var source = new CloudBlockBlob(firstResourceUri);
+            var destination = GetBlockBlobReference(GetName(secondResourceUri));
 
             // For interacting with the source, we just use the same blob request options as the destination blob.
             ApplyBlobRequestOptions(source);
 
-            if (await destination.ExistsAsync())
+            return await AreSynchronized(new AzureCloudBlockBlob(source), new AzureCloudBlockBlob(destination));
+        }
+
+        public async Task<bool> AreSynchronized(ICloudBlockBlob sourceBlockBlob, ICloudBlockBlob destinationBlockBlob)
+        {
+            if (await destinationBlockBlob.ExistsAsync(CancellationToken.None))
```

Comment:
This code can be much more readable if it's restructured a bit:

bool destinationExists = await destinationBlockBlob.ExistsAsync(CancellationToken.None);
bool sourceExists = await sourceBlockBlob.ExistsAsync(CancellationToken.None);

if (!destinationExists || !sourceExists)
{
return destinationExists && sourceExists;
}

 var sourceBlobMetadata = await sourceBlockBlob.GetMetadataAsync(CancellationToken.None);
 var destinationBlobMetadata = await destinationBlockBlob.GetMetadataAsync(CancellationToken.None);

 if (sourceBlobMetadata == null || destinationBlobMetadata == null)
{
       return false;
}

if (!sourceBlobHasSha512Hash)
 {
                        Trace.TraceWarning(string.Format("The source blob ({0}) doesn't have the SHA512 hash.", sourceBlockBlob.Uri.ToString()));
}
if (!destinationBlobHasSha512Hash)
{
                       Trace.TraceWarning(string.Format("The destination blob ({0}) doesn't have the SHA512 hash.", destinationBlockBlob.Uri.ToString()));
  }

if (sourceBlobHasSha512Hash && destinationBlobHasSha512Hash)
{
         if (sourceBlobSha512Hash == destinationBlobSha512Hash)
          {
                   Trace.WriteLine(string.Format("The source blob ({0}) and destination blob ({1}) have the same SHA512 hash and are synchronized.", sourceBlockBlob.Uri.ToString(), destinationBlockBlob.Uri.ToString()));
                    return true;
           }

          Trace.TraceWarning(string.Format("The source blob ({0}) and destination blob ({1}) have the different SHA512 hash and are not synchronized. " +
                            "The source blob hash is {2} while the destination blob hash is {3}",
                            sourceBlockBlob.Uri.ToString(), destinationBlockBlob.Uri.ToString(), sourceBlobSha512Hash, destinationBlobSha512Hash));
                    }
else
{
                    return false;
}

Reply:
This will break the current logic. If destination exists, but source doesn't exist, it should return true. If both of them don't exist, it should return true.
Decide to keep the current one.